Two merged chunks

anne-glerum opened the pull request:
Like two merged boxes, this allows for specifying two different boundary conditions on the lateral boundaries.

bangerth commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,297 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
```
you're living in the past. you ought to be living more in the present!

bangerth commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,297 @@
+     * lowerwest, lowereast, lowersouth, lowernorth,
+     * upperwest, uppereast, uppersouth, uppernorth.
+     * The radius at which the change is boundary indicator occurs,
+     * can be set by the user.
```
grammar

I also don't understand what the sentence is actually trying to say.

bangerth commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,297 @@
+     * The parameters that describe this geometry and that are read from the
+     * input file are the inner and outer radii of the shell, the minimum
+     * and maximum longitude, minimum and maximum longitude, and the
+     * number of cells initialised in each dimension.
```
is this just the same text that was in the "chunk" geometry? If so, you may want to update it (and make the changes suggested above to the original chunk geometry, either in this pull request or a separate one).

anne-glerum replied:
It is based on the chunk text, but extended to describe this plugin (hopefully now in an understandable way).

tjhei commented:
/run-tests

tjhei commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,298 @@
+#include <aspect/geometry_model/interface.h>
+#include <deal.II/grid/manifold.h>
+#include <deal.II/base/function_lib.h>
+#include <deal.II/grid/grid_out.h>
```
do you need this include?

tjhei commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,298 @@
+    using namespace dealii;
+
+    /**
+     * A geometry model class that describes a chunk of a spherical shell.
```
update documentation

tjhei commented:
It looks like you haven't updated the PR with your changes yet. Please add a comment here if we should take another look.

jdannberg commented:
I also have a general comment. It could make sense to derive this model from the chunk model (and only add the things that are new/different compared to the chunk) instead of making a completely new model. This would also have the advantage that in places like in #1044, where you have a different logic depending on the geometry model, you would only need one if-statement for the base model (the chunk) and not one for each of the plugins.

anne-glerum commented:
It's been a while :D, but I've updated this PR to master and incorporated the changes in the chunk geometry. W.r.t. the comment about deriving the TwoMergedChunks from Chunk, I've not done that, because this has been undone for the TwoMergedBoxes in #3488.

anne-glerum commented:
@gassmoeller Noticed this is still open when doing some housekeeping because of the release coming up. No rush from my side tho.

gassmoeller commented on source/geometry_model/two_merged_chunks.cc:
```diff
@@ -0,0 +1,1090 @@
+
+          if (cell->face(3)->at_boundary())
+            // Set the upper part of the eastern boundary to indicator 2*dim+1.
+            if ((cell->vertex(GeometryInfo<dim-1>::vertices_per_cell-1).norm() + cell->vertex(0).norm()) / 2.0 > point3[0])
```
this probably works, but I find it very awkward to read. I think you are looking for the following: `cell->face(3)->center(true)[0] > point3[0]`?

Same below.

gassmoeller commented on source/geometry_model/two_merged_chunks.cc:
```diff
@@ -0,0 +1,1090 @@
+          prm.declare_entry ("Outer radius repetitions", "1",
+                             Patterns::Integer (1),
+                             "Number of cells in radial direction for the upper chunk.");
+          prm.declare_entry ("Grid merge radius repetitions", "1",
```
Would `Inner chunk radius repetitions` and `Outer chunk radius repetitions` be better?

gassmoeller commented on source/geometry_model/two_merged_chunks.cc:
```diff
@@ -0,0 +1,1090 @@
+                                   "chunk with lithosphere boundary indicators",
+                                   "A geometry which can be described as a chunk of a spherical shell, "
+                                   "bounded by lines of longitude, latitude and radius. "
+                                   "The radial boundaries have two boundary indicators, so the user "
```
I find radial is not quite clear, it could mean boundaries of equal radius, or boundaries in radius direction.

```suggestion
                                   "The side boundaries have two boundary indicators, so the user "
```

gassmoeller commented on tests/merged_chunks_2D.prm:
```diff
@@ -0,0 +1,144 @@
+# safely merged with any existing subsections with the same name.
+
+subsection Boundary temperature model
+  set Fixed temperature boundary indicators   = uppereast, upperwest, inner, outer
```
Please replace all `inner` and `outer` by `bottom` and `top`

gassmoeller reviewed (commented):
Thanks for reminding me. I think this looks already good, but I would like to see some changes. Important point:
- If possible we should reuse the chunk manifold instead of adding another manifold
- Replace all use of 'inner' and 'outer' boundaries by 'bottom' and 'top' (see https://github.com/geodynamics/aspect/pull/1744)
- I have some suggestions for renaming parameter, but I am not 100% sure my suggestions are better, feel free to modify otherwise, or let me know if you think the current names are as good as we can get.

gassmoeller commented:
Oh, and we need a changelog entry.

anne-glerum commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,491 @@
+      private:
+        /**
+         * Minimum depth, longitude-depth or
+         * longitude-latitude-depth point. TODO
```
Ha I think I remember why I put TODO there. The comment is copied from chunk.h, but I think in both cases it doesn't make sense list the 1D option (e.g. minimum depth), as we don't do 1D. So the TODO is to update the comment here (which I now did) and in chunk.h (which I will do).

anne-glerum commented on include/aspect/geometry_model/two_merged_chunks.h:
```diff
@@ -0,0 +1,491 @@
+         * The number of cells in each coordinate direction
+         * for the lower and upper chunk.
+         */
+        unsigned int lower_repetitions[dim];
+        unsigned int upper_repetitions[dim];
```
Sure. Will also update for chunk.cc

anne-glerum commented on source/geometry_model/two_merged_chunks.cc:
```diff
@@ -0,0 +1,1090 @@
+      // Create the lower box.
+      GridGenerator::subdivided_hyper_rectangle (lower_coarse_grid,
+                                                 lower_rep_vec,
+                                                 point1,
+                                                 point4,
```
I've added better descriptions to point* in the header file. Is that enough, or do you want them renamed.

anne-glerum commented:
Thanks for the review, I think I've addressed all your comments.

gassmoeller commented:
Thanks, this looks nearly good to go. Please update the test results with the official results. Since you have to do this anyway: Can you maybe decrease the test resolution further by one refinement (or have only have the repetitions for the 3d case)? If that does not work no problem, but I want to keep the test output as small as possible.

anne-glerum commented:
Done. I've reduced the number of repetitions for the 3d chunk and updated the results.

gassmoeller commented:
Thanks, it looks like the output needs another update. And please squash the commits so that we do not merge the intermediate test result files. Ready afterwards.

gassmoeller commented:
Please squash the commits on this branch into one commit to avoid merging the intermediate test results. Afterwards this is ready to go.

anne-glerum commented:
@gassmoeller Squashed and tester passed :)
